zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/MultiResourceDeploymentTest.java

```diff
@@ -0,0 +1,345 @@
+              .isGreaterThan(previousFormKey);
+    }
+
+    private static void assertNewProcessCreatedRecord(final long expectedProcessDefinitionKey) {
+      assertThat(
+              RecordingExporter.processRecords()
+                  .withIntent(ProcessIntent.CREATED)
+                  .limit(2)
+                  .getLast())
+          .satisfies(
+              record -> {
+                assertThat(record.getKey()).isEqualTo(expectedProcessDefinitionKey);
+                assertThat(record.getValue().getProcessDefinitionKey())
+                    .isEqualTo(expectedProcessDefinitionKey);
+                assertThat(record.getValue().getVersion()).isEqualTo(2);
+              });
+    }
+
+    private static void assertNewDecisionRequirementsCreatedRecord(
+        final long expectedDecisionRequirementsKey) {
+      assertThat(
+              RecordingExporter.decisionRequirementsRecords()
+                  .withIntent(DecisionRequirementsIntent.CREATED)
+                  .limit(2)
+                  .getLast())
+          .satisfies(
+              record -> {
+                assertThat(record.getKey()).isEqualTo(expectedDecisionRequirementsKey);
+                assertThat(record.getValue().getDecisionRequirementsKey())
+                    .isEqualTo(expectedDecisionRequirementsKey);
+                assertThat(record.getValue().getDecisionRequirementsVersion()).isEqualTo(2);
+              });
+    }
+
+    private static void assertNewDecisionCreatedRecord(
+        final long expectedDecisionKey, final long expectedDecisionRequirementsKey) {
+      assertThat(
+              RecordingExporter.decisionRecords()
+                  .withIntent(DecisionIntent.CREATED)
+                  .limit(2)
+                  .getLast())
+          .satisfies(
+              record -> {
+                assertThat(record.getKey()).isEqualTo(expectedDecisionKey);
+                assertThat(record.getValue().getDecisionKey()).isEqualTo(expectedDecisionKey);
+                assertThat(record.getValue().getDecisionRequirementsKey())
+                    .isEqualTo(expectedDecisionRequirementsKey);
+                assertThat(record.getValue().getVersion()).isEqualTo(2);
+              });
+    }
+
+    private static void assertNewFormCreatedRecord(final long expectedFormKey) {
+      assertThat(RecordingExporter.formRecords().withIntent(FormIntent.CREATED).limit(2).getLast())
+          .satisfies(
+              record -> {
+                assertThat(record.getKey()).isEqualTo(expectedFormKey);
+                assertThat(record.getValue().getFormKey()).isEqualTo(expectedFormKey);
+                assertThat(record.getValue().getVersion()).isEqualTo(2);
+              });
+    }
+  }
+
+  public static class DeploymentWithNoChangesTest {
+
+    @Rule public final EngineRule engine = EngineRule.singlePartition();
+
+    @Test
+    public void shouldNotCreateNewVersionsIfNoResourceHasChanged() {
```

Comment:
I think this can be more performant, but it's up to you if you like how it looks.
```suggestion
@Test
    public void shouldNotCreateNewVersionsIfNoResourceHasChanged() {
      // given
      final var firstDeployment =
          engine
              .deployment()
              .withXmlResource(PROCESS_V1)
              .withXmlClasspathResource(DMN_V1)
              .withJsonClasspathResource(FORM_V1)
              .deploy();
      final var firstDeploymentValue = firstDeployment.getValue();
      final var processV1 = firstDeploymentValue.getProcessesMetadata().getFirst();
      final var drgV1 = firstDeploymentValue.getDecisionRequirementsMetadata().getFirst();
      final var decisionV1 = firstDeploymentValue.getDecisionsMetadata().getFirst();
      final var formV1 = firstDeploymentValue.getFormMetadata().getFirst();
      final long recordsCountBefore =
          RecordingExporter.records()
              .onlyEvents()
              .limit(l -> l.getPosition() >= firstDeployment.getPosition())
              .count();
      assertThat(recordsCountBefore)
          .isEqualTo(5)
          .describedAs("Should have event records for 4 resources and 1 for deployment created");
      // when
      final Record<DeploymentRecordValue> secondDeployment =
          engine
              .deployment()
              // deploy the exact same resources again
              .withXmlResource(PROCESS_V1)
              .withXmlClasspathResource(DMN_V1)
              .withJsonClasspathResource(FORM_V1)
              .deploy();
      final var secondDeploymentValue = secondDeployment.getValue();
      // then
      assertThat(secondDeploymentValue.getProcessesMetadata())
          .singleElement()
          .satisfies(
              metadata ->
                  Assertions.assertThat(metadata)
                      .hasVersion(1)
                      .isDuplicate()
                      .extracting(
                          ProcessMetadataValue::getProcessDefinitionKey,
                          InstanceOfAssertFactories.LONG)
                      .isEqualTo(processV1.getProcessDefinitionKey()));
      assertThat(secondDeploymentValue.getDecisionRequirementsMetadata())
          .singleElement()
          .satisfies(
              metadata ->
                  Assertions.assertThat(metadata)
                      .hasDecisionRequirementsVersion(1)
                      .isDuplicate()
                      .extracting(
                          DecisionRequirementsMetadataValue::getDecisionRequirementsKey,
                          InstanceOfAssertFactories.LONG)
                      .isEqualTo(drgV1.getDecisionRequirementsKey()));
      assertThat(secondDeploymentValue.getDecisionsMetadata())
          .singleElement()
          .satisfies(
              metadata ->
                  Assertions.assertThat(metadata)
                      .hasVersion(1)
                      .isDuplicate()
                      .extracting(
                          DecisionRecordValue::getDecisionKey,
                          DecisionRecordValue::getDecisionRequirementsKey)
                      .containsExactly(
                          decisionV1.getDecisionKey(), drgV1.getDecisionRequirementsKey()));
      assertThat(secondDeploymentValue.getFormMetadata())
          .singleElement()
          .satisfies(
              metadata ->
                  Assertions.assertThat(metadata)
                      .hasVersion(1)
                      .isDuplicate()
                      .extracting(FormMetadataValue::getFormKey)
                      .isEqualTo(formV1.getFormKey()));

      final long recordsCountAfter =
          RecordingExporter.records()
              .onlyEvents()
              .limit(l -> l.getPosition() >= secondDeployment.getPosition())
              .count();
      assertThat(recordsCountAfter - recordsCountBefore)
          .isEqualTo(1)
          .describedAs("Should only have one extra deployment created");
    }```